requestmanager/asyncloader/asyncloader.go

```diff
@@ -180,20 +221,117 @@ func (al *AsyncLoader) messageQueueWorker() {
+	if existing {
+		return errors.New("already registerd a persistence option with this name")
+	}
+	responseCache, loadAttemptQueue := setupAttemptQueue(rpom.loader, rpom.storer)
+	al.alternateQueues[rpom.name] = alternateQueue{responseCache, loadAttemptQueue}
+	return nil
+}
+
+func (rpom *registerPersistenceOptionMessage) handle(al *AsyncLoader) {
+	err := rpom.register(al)
+	select {
+	case <-al.ctx.Done():
+	case rpom.response <- err:
+	}
+}
+
+func (srm *startRequestMessage) startRequest(al *AsyncLoader) error {
+	if srm.persistenceOption != "" {
+		_, ok := al.alternateQueues[srm.persistenceOption]
+		if !ok {
+			return errors.New("Unknown persistence option")
+		}
+		al.requestQueues[srm.requestID] = srm.persistenceOption
+	}
+	al.activeRequests[srm.requestID] = struct{}{}
+	return nil
 }
 
 func (srm *startRequestMessage) handle(al *AsyncLoader) {
-	al.activeRequests[srm.requestID] = true
+	err := srm.startRequest(al)
+	select {
+	case <-al.ctx.Done():
+	case srm.response <- err:
+	}
 }
 
 func (frm *finishRequestMessage) handle(al *AsyncLoader) {
 	delete(al.activeRequests, frm.requestID)
-	al.loadAttemptQueue.ClearRequest(frm.requestID)
+	loadAttemptQueue := al.getLoadAttemptQueue(al.requestQueues[frm.requestID])
+	loadAttemptQueue.ClearRequest(frm.requestID)
 }
 
 func (nram *newResponsesAvailableMessage) handle(al *AsyncLoader) {
-	al.loadAttemptQueue.RetryLoads()
+	byQueue := make(map[string][]graphsync.RequestID)
+	for requestID := range nram.responses {
+		queue := al.requestQueues[requestID]
+		byQueue[queue] = append(byQueue[queue], requestID)
+	}
+	for queue, requestIDs := range byQueue {
+		loadAttemptQueue := al.getLoadAttemptQueue(queue)
+		responseCache := al.getResponseCache(queue)
+		responses := make(map[graphsync.RequestID]metadata.Metadata, len(requestIDs))
+		for _, requestID := range requestIDs {
+			responses[requestID] = nram.responses[requestID]
+		}
+		responseCache.ProcessResponse(responses, nram.blks)
+		loadAttemptQueue.RetryLoads()
+	}
+}
+
+func (crm *cleanupRequestMessage) handle(al *AsyncLoader) {
+	aq, ok := al.requestQueues[crm.requestID]
+	if ok {
+		al.alternateQueues[aq].responseCache.FinishRequest(crm.requestID)
+		delete(al.requestQueues, crm.requestID)
+		return
+	}
+	al.responseCache.FinishRequest(crm.requestID)
+}
+
+func setupAttemptQueue(loader ipld.Loader, storer ipld.Storer) (*responsecache.ResponseCache, *loadattemptqueue.LoadAttemptQueue) {
+
+	unverifiedBlockStore := unverifiedblockstore.New(storer)
+	responseCache := responsecache.New(unverifiedBlockStore)
+	loadAttemptQueue := loadattemptqueue.New(func(requestID graphsync.RequestID, link ipld.Link) ([]byte, error) {
+		// load from response cache
+		data, err := responseCache.AttemptLoad(requestID, link)
+		if data == nil && err == nil {
+			// fall back to local store
+			stream, loadErr := loader(link, ipld.LinkContext{})
+			if stream != nil && loadErr == nil {
+				localData, loadErr := ioutil.ReadAll(stream)
```

Comment:
This is safe as long as we're sure we trust the local store. Probably fine, just calling this out to make sure.